### ISSUE TITLE
1W TX has no timeout: radio can stay in TRANSMITTER indefinitely if TXDONE never arrives

### ISSUE THREAD

**Issue by beMoD** [danielpetrovic/somfy-io-bridge#5]
Reading the 1W TX path after an unexplained incident at my father's house (details at the end — **I can't prove they're connected**, so please treat that as motivation, not evidence). The code finding stands on its own either way.

Line numbers are from `test-invert`, but this code is identical on `v2026.08.1`.

### There is no way out of TX if TXDONE never arrives

`onTxTicker()` gates everything behind the TXDONE flag (`iohcRadio.cpp:409`):

```cpp
if (!radio->txComplete) {
    ESP_LOGV(TAG, "TX: Waiting for TXDONE... (state=%s)", radioStateToString(radio->radioState));
    return;
}
```

There's no attempt counter, no deadline, and no fallback. The **only** path back to RX is the success path at `iohcRadio.cpp:431-435` (`Sender.detach()` → `packets2send.clear()` → `Radio::setRx()`).

So if TXDONE is never observed — the DIO0 ISR misses it *and* the `REG_IRQFLAGS2` bit-3 fallback at the top of the same function doesn't catch it — the ticker keeps firing and keeps returning early, forever. `REG_OPMODE` was last written to `TRANSMITTER` in `setTx()`, and nothing ever writes it back. **The PA stays keyed indefinitely.**

### Why that failure mode is nastier than a crash

- A continuously keyed PA on 868 MHz makes every io device within range deaf — **including the original Somfy remotes**. From the user's point of view the whole installation is dead, and the one thing they'd normally fall back on doesn't work either.
- The ESP32 itself doesn't hang: WiFi, the API and logging all keep running, the device looks perfectly healthy in HA. So nothing watchdogs it and nothing self-heals. Only cutting power ends it.

### This one isn't inherited from upstream

Worth saying explicitly, because it changes where a fix belongs: cridp's `packetSender()` doesn't wait for TXDONE at all — it drives repeats off the packet's own repeat counter and ends with `stopTransmission()`, so it always terminates. `txComplete` doesn't exist in that file. The TXDONE gate is this repo's own addition, and it is the *more correct* design in principle — waiting for a real confirmation beats blindly clocking on. It just has no escape hatch.

For contrast, the 2W path already does this properly: `TIMEOUT_CMD_MS`, `TIMEOUT_STEP_MS`, `LBT_MAX_RETRIES` and an explicit "transmit anyway after N retries" (`iohc_controller2w.cpp:35-60, 247-253`). The 1W path has only the 500 ms `MIN_INTER_SEND_US` gap between batches.

### Rough shape of a fix

A wall-clock deadline rather than a tick count (`repeatTime` varies per packet), and on expiry: detach the ticker, release the packets from `txCounter` onward (the success path hands ownership to `sent()`, so an abort has to free them itself or it leaks), clear `packets2send`, `Radio::setRx()`, set `RadioState::RX`, log an error, then `startQueuedSend()` so a queued batch isn't stranded.

I'd rather not guess at your intended semantics for `txComplete` (it's also set unconditionally on every DIO0 rising edge in `handle_interrupt_fromisr`, including during RX, which I assume is deliberate), so this is an issue and not a PR — **but I'm happy to write the patch if you tell me what shape you want it in.**

### Separate, much smaller: an unbounded busy-wait inherited from cridp

`SX1276Helpers.h:40`:

```c
#define TxReady  {while (!(readByte(REG_IRQFLAGS1) & RF_IRQFLAGS1_TXREADY));}
```

Used by `setTx()` (`SX1276Helpers.cpp:295-302`) *after* `REG_OPMODE` has already been set to `TRANSMITTER`, so a chip that never raises TxReady spins the CPU with the transmitter on. Same for `RxReady`/`setRx()`. This one is verbatim from `cridp/iown-homecontrol-esp32sx1276` (`include/SX1276Helpers.h:38`), so arguably it belongs upstream — but since the file is vendored here, it's your call whether to bound it locally or leave it. Mentioning it only because it has the same end state as the issue above.

### The incident, for completeness

My father reported that one awning stopped responding to everything, including its original remote. He killed power to the whole house and it worked again afterwards. No logs, no timestamp, and he's a user rather than someone who could tell me whether the io light was affected too — so I'm deliberately not claiming this was the bridge. Motor thermal cutout is the obvious alternative and I can't rule it out. It's simply what made me go looking, and what I found was a code path with no timeout on it.

The bridge has been running at his house since 2026-09-06 with all three devices (two awnings and a dimmer) and this is the only incident so far.

**Comment by danielpetrovic** [comment]
Thanks for tracing this all the way to the register level, and for being upfront that the incident itself is circumstantial, the code finding stands regardless of what actually happened at your father's house.

Confirmed: `onTxTicker()` has no deadline and no attempt counter, only the TXDONE success path returns to RX. Agreed a wall-clock deadline is the right shape rather than a tick count, since `repeatTime` varies per packet.

Planned fix, in two parts since the report itself points at two places with the same shape of hazard:

1. A deadline (proposing 1 second, real frame airtime here is well under 100ms so that's a generous margin) tracked from whenever `Radio::setTx()` is issued, checked in `onTxTicker()` alongside the existing TXDONE check. On expiry: free the remaining queued packets, detach the ticker, force the radio back to RX, and drain anything waiting in the send queue so a stuck batch doesn't strand whatever was queued behind it.
2. Bounding the `TxReady`/`RxReady` busy-waits in `SX1276Helpers.h`/`.cpp` too. A deadline in `onTxTicker()` alone can't reach a CPU stuck inside `setTx()` itself if the underlying PLL/ready flag never appears, the same "keyed PA forever" outcome would still be possible one layer down.

Also found while tracing the packet lifecycle for the abort path: transmitted packets are never actually freed today regardless of this bug, `sent()` only deletes a packet when a TX callback is registered, and the radio is started with `nullptr` for that callback. Unrelated to what was reported here, but fixing it in the same pass since the abort path needs correct packet ownership anyway.

Will put this on a branch and confirm it behaves correctly on the bench (forcing the TXDONE path to fail is easy to simulate) before asking for a real-hardware test, given this one keys the radio if it goes wrong. Will follow up here once there's something to look at.

**Comment by danielpetrovic** [comment]
Folded this into `test-invert`, the branch you already have flashed, since testing this on demand isn't something you can do easily with the hardware at your father's house. No branch switch needed: `external_components` uses `refresh: 0s`, so your next recompile there will pick this up automatically.

What changed: `onTxTicker()` now gives up after a 1 second wall-clock deadline if TXDONE never arrives, instead of waiting forever, and forces the radio back to RX itself. The `TxReady`/`RxReady` busy-waits you flagged separately are bounded the same way. Also fixed a packet-ownership bug found while building the abort path: transmitted packets were never actually being freed (`sent()` only deleted a packet via a TX callback that's never registered).

Compiled clean and running myself now across a full household of shutters, will know by tomorrow morning whether normal operation is unaffected. Nothing needed from you beyond your normal use next time you're there, and flagging it if the original stuck-radio symptom recurs.

**Comment by danielpetrovic** [comment]
[v1.0.0](https://github.com/danielpetrovic/somfy-io-bridge/releases/tag/v1.0.0) is out and includes this fix. `onTxTicker()` now gives up after a 1 second wall-clock deadline if TXDONE never arrives, instead of waiting forever, and forces the radio back to RX itself. The `TxReady`/`RxReady` busy-waits are bounded the same way now too, and the packet-ownership leak found while building the fix is included as well. Nothing needed from you beyond normal use, if the original stuck-radio symptom ever recurs, flag it here.
